scipy/sparse/_coo.py

```diff
@@ -722,16 +778,490 @@ def _matmul_multivector(self, other):
             result_shape = (other.shape[1],)
             col = self.coords[0]
             row = np.zeros_like(col)
-        else:
-            raise NotImplementedError(
-                f"coo_matmat_dense not implemented for ndim={self.ndim}")
-
         result = np.zeros(result_shape, dtype=result_dtype)
         coo_matmat_dense(self.nnz, other.shape[-1], row, col,
                          self.data, other.ravel('C'), result)
         return result.view(type=type(other))
 
 
+    def dot(self, other):
+        if not (issparse(other) or isdense(other) or isscalarlike(other)):
+            # If it's a list or whatever, treat it like an array
+            o_array = np.asanyarray(other)
+
+            if o_array.ndim == 0 and o_array.dtype == np.object_:
+                # Not interpretable as an array; return NotImplemented so that
+                # other's __rmatmul__ can kick in if that's implemented.
+                return NotImplemented
+
+            try:
+                other.shape
+            except AttributeError:
+                other = o_array
+
+        if self.ndim < 3 and (np.isscalar(other) or other.ndim<3):
+            return _spbase.dot(self, other)
+        if isdense(other):
+            return self._dense_dot(other)
+        else:
+            if not (isinstance(self, (coo_array, coo_matrix)) or np.isscalar(self)):
```

Comment:
Use `self.format == "coo"` instead of `isinstance` here and 2 lines later.

Reply:
I realised `dot` in `_coo_base` will only be accessed if `self` has COO format, so I did a bit of refactoring at daa7c79a37924737d3b9b3c045cb04db4794061f.